### ISSUE TITLE
Adopt Go 1.21 to align with Kubernetes

### ISSUE THREAD

**Issue by dusk125** [etcd-io/etcd#16588]
### What would you like to be added?

Kubernetes is updating to Golang 1.21: https://github.com/kubernetes/kubernetes/pull/118996

This issue would bring etcd to be in line with Kubernetes and would keep etcd `main` at the latest Golang version.

In Go 1.21, there are two noteworthy deprecations:
- `math/rand.Read`
- `math/rand.Seed`

Etcd, bbolt, and raft do not use these so this is a simple bump.

Consider updating:

- [x] etcd/main https://github.com/etcd-io/etcd/pull/16594
- [x] ~~etcd/release-3.5~~
- [x] ~~etcd/release-3.4~~
- [x] bbolt/master https://github.com/etcd-io/bbolt/pull/566
- [x] raft/main https://github.com/etcd-io/raft/pull/101

### Why is this needed?

Keep etcd in line with Kubernetes Go version.

**Comment by dusk125** [comment]
I will take assignment of this issue.

**Comment by ahrtr** [comment]
Let's only bump to go 1.21 for main branch, and keep the stable releases as it's for now.

**Comment by ahrtr** [comment]
I recall we discussed this somewhere recently. @jmhbnz @serathius

**Comment by dusk125** [comment]
ack, it was discussed in https://github.com/etcd-io/etcd/issues/16393, I wasn't sure if there was consensus to only update mains with latest or not.

**Comment by jmhbnz** [comment]
Thanks for raising the issue to track this @dusk125.  

Yes we have a [documented policy for golang versions](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/dependency_management.md#golang-versions) now.

**Comment by hasbro17** [comment]
@jmhbnz I'm curious on what the timeline would look like for bumping the Go version for stable branches once they are on an out-of-support Go version? The Go version doc states:
```
... however upgrading minor versions will be avoided unless the minor version in use is now out of support
```

Given that the Go release cadence at every 6 months is faster than etcd's stable branching schedule (e.g 3.5 was around June 2021), I imagine we'll be updating the Go version for 3.5 and other stable branches every time there is a new Go release.

So I'm curious on whether the intended policy is to wait until Go has a new release that makes the current stable branch Go version out of date, and then upgrade?

Not intending to re-litigate the Go version policy and happy to open a new issue if this isn't the right place.

**Comment by jmhbnz** [comment]
> So I'm curious on whether the intended policy is to wait until Go has a new release that makes the current stable branch Go version out of date, and then upgrade?

Thanks for the question. Yes - the intention is stable release branches stay on the same minor go version as long as they are supported. 

As we saw recently with 1.21 being released we had to upgrade both `release-3.4` and `release-3.5` to from go 1.19  to  1.20. 

I'm open to changing the policy in future if something becomes problematic about it.
